cdap-app-fabric/src/main/java/io/cdap/cdap/support/job/SupportBundleJob.java

```diff
@@ -0,0 +1,204 @@
+  private final int maxThreadTimeout;
+  private final Queue<RunningTaskState> runningTaskStateQueue;
+
+  public SupportBundleJob(Set<SupportBundleTaskFactory> supportBundleTaskFactories, ExecutorService executor,
+                          CConfiguration cConf, SupportBundleStatus supportBundleStatus) {
+    this.supportBundleStatus = supportBundleStatus;
+    this.supportBundleTaskFactories = supportBundleTaskFactories;
+    this.supportBundleTasks = new ArrayList<>();
+    this.executor = executor;
+    this.maxRetries = cConf.getInt(Constants.SupportBundle.MAX_RETRY_TIMES);
+    this.maxThreadTimeout = cConf.getInt(SupportBundle.MAX_THREAD_TIMEOUT);
+    this.runningTaskStateQueue = new ConcurrentLinkedQueue<>();
+  }
+
+  /**
+   * parallel processing tasks and generate support bundle
+   */
+  public void generateBundle(SupportBundleTaskConfiguration bundleTaskConfig) {
+    try {
+      File basePath = bundleTaskConfig.getBasePath();
+      supportBundleTasks.addAll(supportBundleTaskFactories.stream().map(factory -> factory.create(bundleTaskConfig))
+                                  .collect(Collectors.toList()));
+      for (SupportBundleTask supportBundleTask : supportBundleTasks) {
+        String className = supportBundleTask.getClass().getName();
+        String taskName = bundleTaskConfig.getUuid().concat(": ").concat(className);
+        SupportBundleTaskStatus taskStatus = initializeTask(taskName, className);
+        executeTask(taskStatus, supportBundleTask, basePath.getPath(), className, taskName, 0);
+      }
+      completeProcessing(basePath.getPath());
+    } catch (Exception e) {
+      LOG.warn("Failed to execute the tasks ", e);
+    }
+  }
+
+  /**
+   * Execute each task to generate support bundle files
+   */
+  public void executeTask(SupportBundleTaskStatus taskStatus, SupportBundleTask supportBundleTask, String basePath,
+                          String className, String taskName, int retryCount) {
+    AtomicLong startTimeStore = new AtomicLong(0L);
+    AtomicReference<SupportBundleTaskStatus> latestTaskStatus = new AtomicReference<>(taskStatus);
+    Future<SupportBundleTaskStatus> futureService = executor.submit(() -> {
+      try {
+        long startTime = System.currentTimeMillis();
+        startTimeStore.set(startTime);
+        latestTaskStatus.set(updateTask(latestTaskStatus.get(), basePath, CollectionState.IN_PROGRESS));
+        supportBundleTask.collect();
+        latestTaskStatus.set(updateTask(latestTaskStatus.get(), basePath, CollectionState.FINISHED));
+      } catch (Exception e) {
+        LOG.warn("Failed to execute task with supportBundleTask {} ", taskName, e);
+        executeTaskAgainAfterFailed(supportBundleTask, className, taskName, latestTaskStatus.get(), basePath,
+                                    retryCount + 1);
+      }
+      return latestTaskStatus.get();
+    });
+    RunningTaskState runningTaskState = new RunningTaskState(futureService, startTimeStore.get(), taskStatus);
```

Comment:
The `startTimeStore.get()` won't work. By the time you do this it would still be 0. If you want to have `RunningTaskState` final, pass the whole AtomicLong and  do `get` when you need the value and check if it's 0